File: typescript/src/part1/chapter4/option.ts
import { match, P } from "ts-pattern";

export type Option<T> = None | Some<T>
export interface None {
    type: 'None'
}
export interface Some<T> {
    type: 'Some'
    value: T
}

export const map = <T, B>(t: Option<T>, f: (t: T) => B): Option<B> =>
    match(t)
        .with({ type: 'None' }, (none) => ({ ...none }))
        .with({ type: 'Some' }, (some) => ({ ...some, value: f(some.value) }))
        .exhaustive()

export const flatMap = <T, B>(t: Option<T>, f: (t: T) => Option<B>): Option<B> =>
    getOrElse(map(t, f), () => ({ type: 'None' }))

// def getOrElse[B >: A](default: => B): B =
export const getOrElse = <T>(t: Option<T>, f: () => T): T =>
    match(t)
        .with({ type: 'None' }, () => f())
        .with({ type: 'Some' }, (some) => some.value)
        .exhaustive()

// def orElse[B >: A](ob: => Option[B]): Option[B] =
export const orElse = <T>(t: Option<T>, f: () => Option<T>): Option<T> =>
    getOrElse(map(t, (x) => ({ type: 'Some', value: x })), f)

export const filter = <T>(t: Option<T>, f: (t: T) => boolean): Option<T> =>
    flatMap(t, (x) => f(x) ? { type: 'Some', value: x } : { type: 'None' } as Option<T>)

const mean = (xs: Array<number>): Option<number> =>
    match(xs)
        .with([], () => ({ type: 'None' } as Option<number>))
        .otherwise(() => ({ type: 'Some', value: xs.reduce((accumulator, x) => accumulator + x, 0.0) / xs.length }))

export const variance = (xs: Array<number>): Option<number> =>
    flatMap(mean(xs), (m) => mean(xs.map((x) => Math.pow(x - m, 2))))

export const map2 = <A, B, C>(a: Option<A>, b: Option<B>, f: (a: A, b: B) => C): Option<C> =>
    flatMap(a, (a) => map(b, (b) => f(a, b)))

export const sequence = <T>(xs: Array<Option<T>>): Option<Array<T>> =>
    match(xs)
        .with([], () => ({ type: 'Some', value: [] } as Option<Array<T>>))
        .otherwise((xs) => {
            const [head, ...tail] = xs
            return flatMap(head, (h) => map(sequence(tail), (t) => [h].concat(t)))
        })

export const traverse = <A, B>(xs: Array<A>, f: (t: A) => Option<B>): Option<Array<B>> =>
    match(xs)
        .with([], () => ({ type: 'Some', value: [] }) as Option<Array<B>>)
        .otherwise((xs) => {
            const [head, ...tail] = xs
            return map2(f(head), traverse(tail, f), (h, t) => [h].concat(t))
        })

export const sequenceFromTraverse = <T>(xs: Array<Option<T>>): Option<Array<T>> =>
    traverse(xs, (x) => x)

File: rust/src/part1/chapter4/option.rs
use std::iter;

#[derive(Debug, Clone, PartialEq)]
enum MyOption<A> {
    MyNone,
    MySome(A),
}

impl<A> MyOption<A>
where
    A: Clone,
{
    fn map<B, F>(&self, f: F) -> MyOption<B>
    where
        F: Fn(A) -> B + Copy,
    {
        match self {
            MyOption::MyNone => MyOption::MyNone,
            MyOption::MySome(value) => MyOption::MySome(f(value.clone())),
        }
    }

    fn flat_map<B, F>(&self, f: F) -> MyOption<B>
    where
        B: Clone,
        F: Fn(A) -> MyOption<B> + Copy,
    {
        self.map(f).get_or_else(|| MyOption::MyNone)
    }

    /// def getOrElse[B >: A](default: => B): B
    fn get_or_else<F>(&self, default: F) -> A
    where
        F: Fn() -> A + Copy,
    {
        match self {
            MyOption::MyNone => default(),
            MyOption::MySome(value) => value.clone(),
        }
    }

    /// def orElse[B >: A](ob: => Option[B]): Option[B]
    fn or_else<F>(&self, ob: F) -> MyOption<A>
    where
        F: Fn() -> MyOption<A> + Copy,
    {
        self.map(|value| MyOption::MySome(value)).get_or_else(ob)
    }

    fn filter<F>(&self, f: F) -> MyOption<A>
    where
        F: Fn(A) -> bool,
    {
        self.flat_map(|value| {
            if f(value.clone()) {
                MyOption::MySome(value)
            } else {
                MyOption::MyNone
            }
        })
    }

    fn map2<B, C, F>(&self, other: MyOption<B>, f: F) -> MyOption<C>
    where
        B: Clone,
        C: Clone,
        F: Fn(A, B) -> C + Copy,
    {
        self.flat_map(|a| other.map(|b| f(a.clone(), b)))
    }
}

fn mean(xs: &[f64]) -> MyOption<f64> {
    if xs.is_empty() {
        MyOption::MyNone
    } else {
        MyOption::MySome(xs.iter().sum::<f64>() / xs.len() as f64)
    }
}

fn variance(xs: &[f64]) -> MyOption<f64> {
    mean(xs).flat_map(|m| mean(&xs.iter().map(|x| (x - m).powf(2.0)).collect::<Vec<_>>()))
}

fn sequence<A>(xs: &[MyOption<A>]) -> MyOption<Vec<A>>
where
    A: Clone,
{
    match xs {
        [] => MyOption::MySome(Vec::<A>::new()),
        [x, xs @ ..] => (*x).flat_map(|head| {
            sequence(xs).map(|tail| iter::once(head.clone()).chain(tail).collect())
        }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_exercise41() {
        //assert_eq!(MyOption::MyNone.map(|x| x + 1), MyOption::MyNone);
        assert_eq!(MyOption::MySome(42).map(|x| x + 1), MyOption::MySome(43));
        //assert_eq!(MyOption::MyNone.flat_map(|x| x + 1), MyOption::MyNone);
        assert_eq!(
            MyOption::MySome(42).flat_map(|x| MyOption::MySome(x + 1)),
            MyOption::MySome(43)
        );
        assert_eq!(MyOption::MyNone.get_or_else(|| 0), 0);
        assert_eq!(MyOption::MySome(42).get_or_else(|| 0), 42);
        assert_eq!(
            MyOption::MyNone.or_else(|| MyOption::MySome(0)),
            MyOption::MySome(0)
        );
        assert_eq!(
            MyOption::MySome(42).or_else(|| MyOption::MySome(0)),
            MyOption::MySome(42)
        );
        //assert_eq!(MyOption::MyNone.filter(|x| <= 42), MyOption::MyNone);
        assert_eq!(
            MyOption::MySome(42).filter(|x| x <= 42),
            MyOption::MySome(42)
        );
    }

    #[test]
    fn test_exercise42() {
        assert_eq!(variance(&[1.0, 2.0, 3.0, 4.0, 5.0]), MyOption::MySome(2.0));
    }

    #[test]
    fn test_exercise43() {
        assert_eq!(
            MyOption::MySome(42).map2::<i32, i32, _>(MyOption::MyNone, |a, b| a + b),
            MyOption::MyNone,
        );
        assert_eq!(
            MyOption::MySome(42).map2::<i32, i32, _>(MyOption::MySome(42), |a, b| a + b),
            MyOption::MySome(84)
        );
    }

    #[test]
    fn test_exercise44() {
        assert_eq!(
            sequence(&[MyOption::MySome(42), MyOption::MyNone]),
            MyOption::MyNone
        );
        assert_eq!(
            sequence(&[MyOption::MySome(42), MyOption::MySome(42)]),
            MyOption::MySome(vec![42, 42])
        );
    }
}
